### PR TITLE
Add 'Complex FFI with Rust & Unity' post.

### PR DESCRIPTION
**Pull request by Schoonology**
I left the original post date so y'all could see when this started, but it's finally ready for review.

There will be some grammatical, spelling, and/or stylistic errors, as I just finished the first, _quick_ edit pass, and wanted some initial feedback before I delve too much further.

- Does this make sense?
- Does it build off the original appropriately?
- Did I miss anything glaring?

As mentioned in the post, the accompanying example project is here: https://github.com/testdouble/rust-ffi-complex-example

### CODE REVIEW

**Comment by rosston** [comment]
For reference, the original is here: http://blog.testdouble.com/posts/2018-01-02-unity-rust-ffi-getting-started

**Comment by Schoonology** [comment]
@stevenjackson Reading back through it (I forgot this wasn't _published_!), I don't see what you're referring to. I do see a section saying that we can use "language features" to guarantee memory _layout_, but that doesn't affect memory consumption, leaks, etc.

Is that what you're asking about?

File: app/posts/2017-11-02-complex-ffi-rust-unity-node.md [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+
+At this point, sadly, there is _no_ guarantee how these are laid out in memory. C# uses the `LayoutKind.Auto` format—[described here][layoutkind]—for classes by default, but explicitly does not make guarantees how memory is used. JavaScript doesn't have a definition, either, leaving it to interpreter. In V8's case, for example, objects are, in the best case, laid out according to "hidden classes"—[described here][v8]—based on which properties have been added.
+
+**TL;DR:** We need to _explicitly_ define how the langauges we use—both the "host" language and the "embedded" language—lay out shared structures in memory. Once we do, we should be able to use _native language features_ to access these structures on both sides of the bridge.
+
+To make this easier, the [C programming language][c] is, for one reason or another, the [lingua franca][lingua-franca] of memory layouts for languages. So much so that we can often tell other languages "just use C's rules", and get the consistency we require.
+
+### C-style in C#: LayoutKind.Sequential
+
+To get a "c-style" class in C#, apply the `LayoutKind.Sequential` struct layout to the class:
```
Comment (rosston) [inline app/posts/2017-11-02-complex-ffi-rust-unity-node.md]: Should this be `"C-style"`?

File: app/posts/2017-11-02-complex-ffi-rust-unity-node.md [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+
+```csharp
+[StructLayout(LayoutKind.Sequential)]
+public class Example {
+  ...
+```
+
+### C-style in Rust: repr(C)
+
+To get a "c-style" data type in Rust, apply the `C` representation to the struct or enum:
```
Comment (rosston) [inline app/posts/2017-11-02-complex-ffi-rust-unity-node.md]: Should this be `"C-style"`?

File: app/posts/2017-11-02-complex-ffi-rust-unity-node.md [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+  ...
+```
+
+### Also, strings
+
+While we won't go through every possible higher-level language feature and how to fit it through an FFI bridge, let's look at strings. Strings are often considered "primitives" in languages like JavaScript and C#, but that's unfortunately naïve. A string is a collection type, just like an array, though it is a nicely strict collection: it only contains characters.
+
+Even the "simple" collection presents the same problem: how big are these characters? Are they stored as [US-ASCII][us-ascii], [ISO 8859][iso-8859], or some [Unicode encoding][unicode]?
+
+Just as we did with objects, we'll need to indicate _explictly_ how to treat strings. Fortunately for our example, Rust is already specific about it's standard `String` type: [it is UTF-8 encoded][rust-utf-8]. The only work we need to do, then, is be explicit on the C# side (sample taken from the example project):
```
Comment (rosston) [inline app/posts/2017-11-02-complex-ffi-rust-unity-node.md]: >about ~it's~ its standard `String` type

File: app/posts/2017-11-02-complex-ffi-rust-unity-node.md [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+
+While some langauges (:cough: C :cough:) play fast-and-loose with the program's (and programmer's) ability to leak memory, most modern languages are more restrictive with the lifespan of any allocated block of memory. What this means for our FFI work is that memory allocated in Rust, for example, is _owned_ by the Rust library, and it's the Rust library's job to ensure that memory is preserved. If a pointer to that block of memory is passed to C#, and Rust allows that memory to be freed, any access to that memory from C# will result in a segmentation fault and a crash.
+
+Each set of languages and libraries you use will present a different set of options, so I'll leave you with the following general advice: if one your project's languages makes explicit memory management (e.g. allocating, freeing, pointer math) _more difficult_, that language should own any shared memory. In our example, that langauge is C#: it is much easier to use the `unsafe` block in Rust and do whatever pointer trickery we require, so we'll allocate objects in C# that are written to in Rust, rather than trying to pass Rust-allocated memory to C#.
```
Comment (stevenjackson) [inline app/posts/2017-11-02-complex-ffi-rust-unity-node.md]: @Schoonology This was the advice I perked my ears up at.  I initially read it as something like: "Pick the stricter language for allocating for safety reasons".  Which made me paranoid - I feel like you should have some sense of "memory leaks likely here" before you start taking on shared allocation problems.

On my latest read, I see it's something more like - managing memory in C# is harder, so let's not do it there.  So maybe there's a way to improve that and make it clearer, or maybe I'm just looking for nits to pick.

**Comment by Schoonology** [comment]
I made a couple smaller changes, and looked at the section @stevenjackson mentioned a second time, but was unable to come up with a better way to express it without a more significant rewrite.

Barring objection, I'll merge this later today.

**Comment by kbaribeau** [comment]
💯 I enjoyed the read, even though I just read the diff on GitHub. It would've been even better on the real blog :)

**Comment by rosston** [comment]
@Schoonology Before merge, might want to look into why the review app isn't deploying. It seems like other recent PRs are deploying fine, so I wonder if somehow this is failing due to something your post (even though that seems unlikely).

**Comment by searls** [comment]
^ If you rebase, it ought to deploy successfully. Yesterday I fixed an issue in testdouble/style that was preventing builds

**Pull request merged**